pkg/auth/providers/azuread/azuread_client.go

```diff
@@ -0,0 +1,550 @@
+	}
+
+	resp, err := ac.postToAzureAD(url, body.String())
+	if err != nil {
+		logrus.Errorf("AzureAD getAccessToken: GET url %v received error from Azure, err: %v", url, err)
+		return "", "", err
+	}
+	defer resp.Body.Close()
+
+	// Decode the response
+	var respMap map[string]interface{}
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		logrus.Errorf("AzureAD getAccessToken: received error reading response body, err: %v", err)
+		return "", "", err
+	}
+
+	if err := json.Unmarshal(b, &respMap); err != nil {
+		logrus.Errorf("AzureAD getAccessToken: received error unmarshalling response body, err: %v", err)
+		return "", "", err
+	}
+
+	if respMap["error"] != nil {
+		desc := respMap["error_description"]
+		logrus.Errorf("Received Error from AzureAD %v, description from AzureAD %v", respMap["error"], desc)
+		return "", "", fmt.Errorf("Received Error from AzureAD %v, description from AzureAD %v", respMap["error"], desc)
+	}
+
+	accessToken, ok := respMap["access_token"].(string)
+	if !ok {
+		return "", "", fmt.Errorf("Received Error reading accessToken from response %v", respMap)
+	}
+
+	refreshToken, ok := respMap["refresh_token"].(string)
+	if !ok {
+		return "", "", fmt.Errorf("Received Error reading refreshToken from response %v", respMap)
+	}
+	return accessToken, refreshToken, nil
+}
+
+func (ac *Client) postToAzureAD(url string, body string) (*http.Response, error) {
+	req, err := http.NewRequest("POST", url, strings.NewReader(body))
+	if err != nil {
+		logrus.Error(err)
+	}
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := ac.httpClient.Do(req)
+	if err != nil {
+		logrus.Errorf("Received error from azure: %v", err)
+		return resp, err
+	}
+	// Check the status code
+	switch resp.StatusCode {
+	case 200:
+	case 201:
+	default:
+		var body bytes.Buffer
+		io.Copy(&body, resp.Body)
+		return resp, fmt.Errorf("Request failed, got status code: %d. Response: %s",
+			resp.StatusCode, body.Bytes())
+	}
+	return resp, nil
+}
+
+func (ac *Client) getUser(azureAccessToken string, config *v3.AzureADConfig) (Account, map[string]string, error) {
+	if azureAccessToken == "" {
+		return Account{}, nil, httperror.NewAPIError(httperror.ServerError, "No Azure Access token")
+	}
+
+	authority, graphEndpoint, _, err := ac.uriEndpointConstruction(config)
+	if err != nil {
+		logrus.Errorf("Wrong Azure Domain %v is provided", config.Domain)
+		return Account{}, nil, fmt.Errorf("Wrong Azure Domain %v is provided", config.Domain)
+	}
+	url, err := ac.getURL(authority, graphEndpoint, "USER", config, "")
+
+	resp, newAccessToken, providerInfo, err := ac.getFromAzureAD(azureAccessToken, config, url)
+	if err != nil {
+		logrus.Errorf("AzureAD getAzureADUser: GET url %v received error from Azure, err: %v", url, err)
+		return Account{}, nil, err
+	}
+	if newAccessToken != "" {
+		resp, _, providerInfo, err = ac.getFromAzureAD(newAccessToken, config, url)
```

Comment:
are you going to replace the new, non-empty providerInfo that you got from line 172 with an empty one here?